core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,143 @@
+                    request.getHostGroupName(),
+                    collectStartedInstancesFromCloudProvider(connector, ac, request.getAllInstancesInHostGroup())
+                            .stream()
+                            .map(CloudInstance::getInstanceId).collect(Collectors.toSet()));
+
+            ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+
+            DetailedHostStatuses detailedHostStatuses = clusterHealthService.getDetailedHostStatuses();
+
+            List<InstanceMetadataView> instanceMetadataWithServicesNotRunning = collectRecoveryCandidates(startedInstances, detailedHostStatuses);
+
+            List<CloudInstance> instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(instanceMetadataWithServicesNotRunning,
+                    stack.getStack());
+
+            StopStartUpscaleGetRecoveryCandidatesResult result =
+                    new StopStartUpscaleGetRecoveryCandidatesResult(request.getResourceId(), request, instancesWithServicesNotRunning,
+                            request.getAdjustment(), request.getHostGroupName());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect upscale recovery candidates for stack: %s, proceeding with subsequent flow.",
+                    request.getResourceId());
+            LOGGER.error(errorMessage, e);
+            StopStartUpscaleGetRecoveryCandidatesResult result = new StopStartUpscaleGetRecoveryCandidatesResult(request.getResourceId(), request,
+                    Collections.emptyList(), request.getAdjustment(), request.getHostGroupName());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector<?> connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    private AuthenticatedContext getAuthenticatedContext(StopStartUpscaleGetRecoveryCandidatesRequest request, CloudContext cloudContext,
+            CloudConnector<?> connector) {
+        return connector.authentication().authenticate(cloudContext, request.getCloudCredential());
+    }
+
+    private List<InstanceMetadataView> collectRecoveryCandidates(List<InstanceMetadataView> startedInstances, DetailedHostStatuses detailedHostStatuses) {
+        Predicate<InstanceMetadataView> instanceHealthCheck = im -> {
+            HostName hostName = hostName(im.getDiscoveryFQDN());
+            return detailedHostStatuses.isHostDecommissioned(hostName)
```

Comment:
Can these checks be very explicit to map to the scenarios identified in the writeup. That will lead to some repetitive  checks, but is much easier to read and ensure that the scenarios are covered.

host healthy + Services not running is a valid recovery scenario - which I don't think is covered by this check.

Reply:
Ack, will modify these predicates to correspond with the ones present in the doc, and also split the host and services healthChecks

Reply:
Updated this to correspond with the write up